docs/source/advanced_usage.rst

```diff
@@ -425,6 +425,141 @@ We can run this program a few times to see what we get in the readout register `
      [1]
      [0]]
 
+Sentinel based loop
+===================
+
+Now that we understand how to create loops and conditionals, we can put them together to create a sentinel controlled
+loop. That is, we'll repeat the body of a program until a certain condition is met. In this example, we'll use the
+classic bell state program to demonstrate the concept. However, this technique can be applied to any program with a
+probabilistic outcome that we want to repeat until we get a desired result.
+
+To start, let's import everything we'll need:
+
+.. testcode:: sentinel-based-loop
+
+    # Import some types we'll use
+    from typing import Optional, Tuple
+
+    # We'll use numpy to help us validate our results
+    import numpy as np
+
+    # We'll need to create a program and define an executor
+    from pyquil import Program, get_qc
+    # We'll use these gates in our program
+    from pyquil.gates import CNOT, H, X
+    # We'll also need the help of a few control flow instructions
+    from pyquil.quilbase import Halt, Qubit, MemoryReference, JumpTarget, Jump
+    from pyquil.quilatom import Label
+
+Building our program
+--------------------
+
+Adding control flow to a program introduces complexity, especially as we add more branches to the program. To manage 
+this complexity we'll use some of the methods we learned about in the previous sections as well as by breaking down the
+program into its constituent parts.
+
+The program body
+^^^^^^^^^^^^^^^^
+
+First, let's define the body of our program. This is the part of the program that we'll repeat until we get the result
+we desire. In this case, we'll create a bell state between two qubits and measure them:
+
+.. testcode:: sentinel-based-loop
+
+    def body(qubits: Tuple[Qubit, Qubit], measures: MemoryReference) -> Program:
+        """Constructs a bell state between the given qubit and measures them into the given memory reference."""
+        program = Program(H(qubits[0]), CNOT(*qubits))
+        program.measure_all(*zip(qubits, measures))
+        return program
+
+Resetting state
+^^^^^^^^^^^^^^^
+
+For this program, we'll say our desired result is that both qubits measure to 0. After an unsuccessful attempt where
+they measure to 1, we'll want to reset the state of the qubits before trying again. To do this, we'll create a program
+that applies an :math:`X` gate to the qubits if either of them measured to 1:
+
+.. testcode:: sentinel-based-loop
+
+    def reset_bell_state(qubits: Tuple[Qubit, Qubit], measures: MemoryReference) -> Program:
+        """Resets the state of the qubits if either of them measured to 1."""
+        program = Program()
+        program.if_then(measures[0], Program(X(qubits[0]), X(qubits[1])))
```

Comment:
did you meant to reset these independently? On noiseless QVM they'll always be equal but not on QPU

Reply:
Oops yeah, the initial draft had that in it. At some point I took a frictionless sphere approach to the guide and removed it, but I think it's best left in for correctness.